cloud_shadow_cover.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np # maths 
+import matplotlib.pyplot as plt # plots
+import cv2 # image processing
+import rasterio # file loading
+
+class ImageMasker():
+
+    '''
+    Loads and masks a satellite image. 
+    Arguments:
+    path_to_image: the path to the image, .tif format
+    path_to_land_mask: path to the land mask, .npy format 
+    '''
+    
+    def __init__(self, path_to_image, path_to_land_mask):
+        # Assign paths
+        self.path_to_image = path_to_image
+        self.path_to_land_mask = path_to_land_mask
+        
+        # Set Landsat / Sentinel 2 scaling
+        if 'LS' in self.path_to_image:
+            self.scale = 66000
+        else:
+            self.scale = 10000
+        
+        # Load data
+        self.load_data()
+        
+        # Get cloud and shadow masks
+        self.cloud_mask = self.get_cloud_mask()
+        self.shadow_mask = self.get_shadow_mask()
+        
+        # Add cloud and shadow masks
+        self.full_mask = np.logical_or(self.cloud_mask, self.shadow_mask)
+        
+        # Apply masks to images
+        self.cloud_masked_image = self.apply_mask(self.image, self.cloud_mask)
+        self.shadow_masked_image = self.apply_mask(self.image, self.shadow_mask)
+        self.full_masked_image = self.apply_mask(self.image, self.full_mask)
+        self.land_full_masked_image = self.apply_mask(self.full_masked_image, self.land_mask)
+
+        # Calculate cover percentages
+        self.ocean_cloud_cover_percent = self.ocean_cover_percent(self.cloud_mask)
+        self.ocean_shadow_cover_percent = self.ocean_cover_percent(self.shadow_mask)
+        self.ocean_full_cover_percent = self.ocean_cover_percent(self.full_mask)
+        
+    def load_data(self):
+        '''
+        Loads image, cloud bits and land mask
+        '''
+        # Load image
+        image = rasterio.open(self.path_to_image).read()
+        # Scale image
+        self.image = np.dstack([image[2]/self.scale, image[1]/self.scale, image[0]/self.scale])*3
+        # Load cloud bits
+        self.cloud_data = rasterio.open(self.path_to_image.replace('rgb', 'extra')).read(4)
+        # Load land mask
+        mask = np.load(self.path_to_land_mask)
+        mask = mask.astype(np.uint8)
+        # Resize land mask
+        self.land_mask = cv2.resize(mask, self.image.shape[:2][::-1]) != 1
+      
+    def get_cloud_mask(self):
+        '''
+        Gets a cloud mask from the cloud data
+        '''
+        # Convert cloud data from int to binary
+        cloud_bits = self.vec_bin_array(self.cloud_data, 16)
+        # Create logical array as a mask
+        cloud_mask = cloud_bits[:, :, 12] == 1
```

Comment:
I think it might be worth adding a comment for this line that says that the 12th index is actually the 3rd bit. Same goes for the shadow mask on line 80, 11th index is actually 4th bit.

Reply:
Have updated the file accordingly!